src/main/java/com/google/cloud/spanner/pgadapter/statements/CopyStatement.java

```diff
@@ -0,0 +1,203 @@
+  }
+
+  public MutationWriter getMutationWriter() {
+    return this.mutationWriter;
+  }
+
+  private void verifyCopyColumns() throws SQLException {
+    if (options.getColumnNames().size() > this.tableColumns.size()) {
+      throw new SQLException("Number of copy columns provided exceed table column count");
+    }
+    LinkedHashMap<String, TypeCode> tempTableColumns = new LinkedHashMap<>();
+    // Verify that every copy column given is a valid table column name
+    for (String copyColumn : options.getColumnNames()) {
+      if (!this.tableColumns.containsKey(copyColumn)) {
+        throw new SQLException(
+            "Column \"" + copyColumn + "\" of relation \"" + getTableName() + "\" does not exist");
+      }
+      // Update table column ordering with the copy column ordering
+      tempTableColumns.put(copyColumn, tableColumns.get(copyColumn));
+    }
+    this.tableColumns = tempTableColumns;
+  }
+
+  private static TypeCode parseSpannerDataType(String columnType) {
+    // Eliminate size modifiers in column type (e.g. STRING(100), BYTES(50), etc.)
+    int index = columnType.indexOf("(");
+    columnType = (index > 0) ? columnType.substring(0, index) : columnType;
+    TypeCode type = TypeCode.valueOf(columnType);
+    if (type == null) {
+      throw new IllegalArgumentException(
+          "Unrecognized or unsupported column data type: " + columnType);
+    }
+    return type;
+  }
+
+  private void queryInformationSchema() throws SQLException {
+    Map<String, TypeCode> tableColumns = new LinkedHashMap<>();
+    PreparedStatement statement =
+        this.connection.prepareStatement(
+            "SELECT "
+                + COLUMN_NAME
+                + ", "
+                + SPANNER_TYPE
+                + " FROM information_schema.columns WHERE table_name = ?");
+    statement.setString(1, getTableName());
+    ResultSet result = statement.executeQuery();
+
+    while (result.next()) {
+      String columnName = result.getString(COLUMN_NAME);
+      TypeCode type = parseSpannerDataType(result.getString(SPANNER_TYPE));
+      tableColumns.put(columnName, type);
+    }
+
+    if (tableColumns.isEmpty()) {
+      throw new SQLException("Table " + getTableName() + " is not found in information_schema");
+    }
+
+    this.tableColumns = tableColumns;
+
+    if (options.getColumnNames() != null) {
+      verifyCopyColumns();
+    }
+  }
+
+  @Override
+  public void execute() {
+    this.executed = true;
+    try {
+      parseCopyStatement();
+      mutationWriter =
+          new MutationWriter(
+              options.getTableName(), this.tableColumns, getParserFormat(), hasHeader());
+      updateResultCount(); // Gets update count, set hasMoreData false and statement result
+    } catch (Exception e) {
+      SQLException se = new SQLException(e.toString());
+      handleExecutionException(se);
+    }
+  }
+
+  private void parseCopyStatement() throws Exception {
+    try {
+      parse(sql, this.options);
```

Comment:
can you turn the getParserFormat function into a setParserFormat function that is called here and sets the resulting value on this.format? It doesn't look like that member variable is being used at all right now. And then getParserFormat will just return the member variable